### ISSUE TITLE
Integrating ARKIT and ARCORE with flutter

### ISSUE THREAD

**Issue by khal-it** [flutter/flutter#22860]
Hello awesome people! 
First of all I want to thank the Flutter team for making our lives sooooo....*one eternity later*.....oooooo much easier! 
I currently started working on an App using Flutter and love it!!!!! 
So recently I got a job offer as an IOS/Android App developer and i'm trying to convince the company to use Flutter for developing the App. There is only one problem though. I have to find a way to integrate augmented reality via ARKIT and ARCORE. I know Flutter is currently only 2D.  And I  don't have much experience in the field of augmented reality, but i thought maybe there is a way to solve this problem. 

Thank you in advance! 
👍

**Comment by neokree** [comment]
I am not a contributor (yet) , but I think that the ARKit & ARcore could be made available developing a "plugin package" and wrapping their functionality in a common dart interface. 
Look [here](https://flutter.io/developing-packages/) if you are interested

**Comment by rhokstar** [comment]
I'm really looking forward to this!

**Comment by Wisdom0063** [comment]
I seriously need this feature

**Comment by JosephSaw** [comment]
Currently working on a project that requires AR to be implemented. This feature would truly help!

**Comment by knightcube** [comment]
Waiting for this for pretty long now

**Comment by khalithartmann** [comment]
I Implemented a plugin that recognizes images via ARCore. will publish it soon . Would be cool if you guys could help me  implement new features and improve the plugin. I also want the plugin to support ARKit. Will here from me soon!

**Comment by knightcube** [comment]
@khalithartmann where can I find the repo?

**Comment by hellowtisch** [comment]
@khalithartmann: I also am interested in your code

**Comment by khalithartmann** [comment]
will upload the plugin soon. will let you guys know

**Comment by JosephSaw** [comment]
Hey, @khalithartmann thank you for taking the time in creating a plugin to integrate ARCore with Flutter.

I will be having a very important meeting early next month where I would need to demonstrate an AR prototype for a client. I would really love to use Flutter as the UI Creation and Experience is simply, amazing. 

So I was just wondering if there was anything I could help out with that could potentially speed up your release? Otherwise, I might have to look into other options for this prototype.

**Comment by hellowtisch** [comment]
I'd also be happy to help, as we also face the decision, whether we can use flutter.




Outlook für Android herunterladen







On Mon, Mar 11, 2019 at 3:07 PM +0100, "JosephSaw" <notifications@github.com> wrote:












Hey, @khalithartmann thank you for taking the time in creating a plugin to integrate ARCore with Flutter.


I will be having a very important meeting early next month where I would need to demonstrate an AR prototype for a client. I would really love to use Flutter as the UI Creation and Experience is simply, amazing.


So I was just wondering if there was anything I could help out with that could potentially speed up your release? Otherwise, I might have to look into other options for this prototype.

**Comment by khalithartmann** [comment]
I just  published the plugin. Help is always appreciated!
https://pub.dartlang.org/packages/arcore_plugin/versions/0.0.1

**Comment by knightcube** [comment]
Need some help integrating it @khalithartmann

**Comment by khalithartmann** [comment]
hey @knightcube , where are you facing the problems ?

**Comment by knightcube** [comment]
I wanted to understand how to load the images in the Augmented Database in Flutter. I think raising an issue in the plugin's repo would be good

**Comment by khalithartmann** [comment]
You just have to create a file in a temp directory : 
`
final Directory systemTempDir = Directory.systemTemp;
    final File tempFile = File('${systemTempDir.path}/image_database.imgdb');

    // create tempfile
    await tempFile.create();
`
and then you have to save the content of the imgdb into the temp file. 
I will add this to the documentation tomorrow. 
In the future it will be possible to pass the path of the imgdb file as an argument to the ArCoreView
Hope i could help.

**Comment by khalithartmann** [comment]
Ah and as for now you cant train the IMGDB in the app. you have to pretrain it with the arcoreimg tool. see  https://developers.google.com/ar/develop/c/augmented-images/arcoreimg

**Comment by hellowtisch** [comment]
Does the plugin and example run under ios? 

I get lots of errors, where I am not sure, whether this is due to config errors on my side, or due to missing ios implementation?

 

Thanks

 

Von: khalithartmann <notifications@github.com>
Antworten an: flutter/flutter <reply@reply.github.com>
Datum: Freitag, 8. März 2019 um 22:20
An: flutter/flutter <flutter@noreply.github.com>
Cc: hellowtisch <marcel@hellofaservice.de>, Comment <comment@noreply.github.com>
Betreff: Re: [flutter/flutter] Integrating ARKIT and ARCORE with flutter (#22860)

 

will upload the plugin soon. will let you guys know

**Comment by khalithartmann** [comment]
Hey no the plugin does not support IOS because ARCore only supports android

**Comment by hellowtisch** [comment]
Hey arcore also Supports iOS, according to the specs, that's why your plugin is so important




Outlook für Android herunterladen







On Mon, Mar 18, 2019 at 10:11 AM +0100, "khalithartmann" <notifications@github.com> wrote:












Hey no the plugin does not support IOS because ARCore only supports android

**Comment by khalithartmann** [comment]
Currently only the cloud anchors functionality is supported for IOS. The Image recognition functionality is not supported as stated in the documentation. Could you explain to me what you mean by "Hey arcore also Supports iOS, according to the specs" ? Do you have a link or something like that ? Maybe you could help adding the IOS support for the plugin if its possible

**Comment by jakoby12** [comment]
Hi!

You can find a ARKit puglin!
https://pub.dartlang.org/packages/arkit_plugin#-changelog-tab-

And the repo
https://github.com/olexale/arkit_flutter_plugin

**Comment by gvt** [comment]
+1 for support for an isomorphic AR solution for Flutter with iOS + Android. Seems like integrating ARCore for Android and ARKit for iOS would be a winning combination. Maybe a new plugin will be born to have ARKit and ARCore combined.

**Comment by tonyshkurenko** [comment]
Would be nice, also waiting for that.

**Comment by penguin3069** [comment]
wait for this feature.

**Comment by Jordan-Gillard** [comment]
A single plugin for both ARCore and ARKit would be incredible.

**Comment by Harbdollar** [comment]
Thanks for the plugin, I really love it. I am curious about **ARcore** and **ARkit**, What is the different between the two? Can I use both on the same project?

 THANKS

**Comment by umar0645** [comment]
> 
> 
> Hi!
> 
> You can find a ARKit puglin!
> https://pub.dartlang.org/packages/arkit_plugin#-changelog-tab-
> 
> And the repo
> https://github.com/olexale/arkit_flutter_plugin

Is it just for ios or is it also support android?

**Comment by MarcoYMD** [comment]
Any update on this? I am really looking forward for a standard solution for flutter

**Comment by gangov** [comment]
hey, if we can get a single plugin for both ARCore and ARKit, that'd be great, you have my vote on this one!

**Comment by CariusLars** [comment]
Hey everyone,
I stumbled upon this thread a while ago when working on an AR project for a client that wanted the new AR functionality embedded into their existing Flutter app. We didn't end up building the app in Flutter back then, but the idea of a cross-platform AR plugin for flutter has bugged me ever since, so I've decided to dedicate my Master's Thesis to the creation of it.
My current plan is combining the [arkit_plugin](https://pub.dev/packages/arkit_plugin) and [arcore_plugin](https://pub.dev/packages/arcore_plugin) into one single plugin (featuring all the new stuff such as Google Cloud Anchors), as @gvt, @MarcoYMD, @gangov and others suggested. I already have a few ideas to make this plugin intuitive and easy-to-use, but seeing how active this community is I thought you might have some suggestions and wishes as well so I end up building something useful for all of you.
So, long story short, please comment everything that you'd wish for in the plugin!

**Comment by gangov** [comment]
hey, @CariusLars thank you for looking up to this, I personally don't have any specific requests for the plugin, making one common place for both the arkit_plugin and arcore_plugin is more than enough. I hope we can contribute as well if you need help.

cheers,

**Comment by RobinLbt** [comment]
@CariusLars That would be a very useful plugin ! I think geolocalisation would be a good main feature. 
(Getting Lat and Long position and place them in the view)

**Comment by CariusLars** [comment]
Good idea @RobinLbt , I'll put it on the list! What exactly do you mean with "place them in the view"? Just attaching the information to the object when uploading it and then being able to run geo-based queries or do you mean something else?

**Comment by RobinLbt** [comment]
Sorry my explanation was pretty bad haha, I meant being able to place GPS coordinates (Lat and long) in the AR coordinates system without having to do all the calculation, this would be very useful since a lot of AR projects are using geographical data

**Comment by Tragio** [comment]
> Hey everyone,
> I stumbled upon this thread a while ago when working on an AR project for a client that wanted the new AR functionality embedded into their existing Flutter app. We didn't end up building the app in Flutter back then, but the idea of a cross-platform AR plugin for flutter has bugged me ever since, so I've decided to dedicate my Master's Thesis to the creation of it.
> My current plan is combining the [arkit_plugin](https://pub.dev/packages/arkit_plugin) and [arcore_plugin](https://pub.dev/packages/arcore_plugin) into one single plugin (featuring all the new stuff such as Google Cloud Anchors), as @gvt, @MarcoYMD, @gangov and others suggested. I already have a few ideas to make this plugin intuitive and easy-to-use, but seeing how active this community is I thought you might have some suggestions and wishes as well so I end up building something useful for all of you.
> So, long story short, please comment everything that you'd wish for in the plugin!

I was researching and I was surprised how this was not still a thing in React or Flutter. Would love to see this happening! 🥳

**Comment by RobinLbt** [comment]
Hello @gangov ,

After using both plugins for a project, the flutter plugin is much more advanced than the ARCore plugin, where a lot of native features of ARCore are not implemented yet.
Maybe you should try to start from 0 instead of building it on top of ARCore plugin for Flutter.
 
Location-based AR (GPS coords) is very popular, and it would be a very useful feature if it could be implemented in your plugin !
I've made all the functions for calculation to do it manually if you are interested. 

Also, a feature to make objects always looking at the camera would be useful.

As for the API, ARKit and ARCore share similar principles so it might be possible to create universal methods for both.
For specific features of each, you could have separated methods that are plateform specific if universal ones is not possible

Do you have any ideas for the name of the plugin ? **AR_plugin** is clear and simple !

**Comment by CariusLars** [comment]
> Hello @CariusLars,
> 
> After using both plugins for a project, the ARKit flutter plugin is much more advanced than the ARCore plugin, where a lot of native features of ARCore are not implemented yet.
> Maybe you should try to start from 0 instead of building it on top of ARCore plugin for Flutter.
> 
> Location-based AR (GPS coords) is very popular, and it would be a very useful feature if it could be implemented in your plugin !
> I've made all the functions for calculation to do it manually if you are interested.
> 
> Also, a feature to make objects always looking at the camera would be useful.
> 
> As for the API, ARKit and ARCore share similar principles so it might be possible to create universal methods for both.
> For specific features of each, you could have separated methods that are plateform specific if universal ones is not possible
> 
> Do you have any ideas for the name of the plugin ? **AR_plugin** is clear and simple !

Hi @RobinLbt,
thanks for your input! Good to know that ARCore plugin is lagging behind by that much, I've only used parts of it so wasn't aware of the overall state. 

Regarding the features you suggested: Location-based AR is definitely on the roadmap as a must-have feature. I've done something similar before in Kotlin, but would appreciate having a look at your approach! Is the code in a repo you could add me to with viewing permissions?
Objects always looking at the camera seems a bit more niche, but could be a cool feature so I've noted it down a bit further down the line.

As an update to the overall progress: I've been busy doing literature research and all that kind of stuff for my thesis but will start architecture design for the plugin next week, so hopefully I can set up the repo soon and start with some coding. I can also share a few options for the API architecture if people are interested in discussing. **AR_plugin** sounds like a good name to me. 
I'll keep you guys posted, I'm very happy about all the feedback I've got so far!

**Comment by RobinLbt** [comment]
Very excited for this project !

Let me know when the repo is created, I will add the location-based functions afterward (still need some testing and code cleaning)

**Comment by CariusLars** [comment]
> Very excited for this project !
> 
> Let me know when the repo is created, I will add the location-based functions afterward (still need some testing and code cleaning)

Will do, sounds like a plan!

**Comment by CariusLars** [comment]
Finally got around to creating the repo: https://github.com/CariusLars/ar_flutter_plugin
In the next 2-3 weeks I'll set up the coarse structure in the code, but I've already laid down a few thoughts about the architecture in the readme (and the example app readme) so feel free to comment on that.
My main focus in the beginning will be collaborative AR, so making sure people can easily use the plugin for creating experiences across multiple devices. Features like image anchors, light estimation and so on aren't something I'll look into right away, but I think those are parts people might want to contribute once I've set up the general code structure (because development can happen in parallel) and a lot can also be taken from the existing ARKit Plugin and the ARCore plugin.

**Comment by Marionetista** [comment]
Hey guys, I want to start a new AR Flutter project. I need to create a little MVP in two weeks for my client so I can't just wait the @CariusLars's plugin to be finished until there hehe. 

You guys recommend using ARKit and ARCore plugins in the same project? Does that consumes a lot of storage capacity? Anyone have a project to share, just to understand how to manage both plugins inside the same project? I'll be very glad.

I've seen a lot of people using https://pub.dev/packages/flutter_unity_widget to create simple AR apps, but I don't know how to use the Unity... so I'm a bit afraid. Does the AR on the Unity Plugin compatible for IOS and Android??

When the ar_flutter_plugin is ready, I will be, certainly using it!

**Comment by CariusLars** [comment]
> Hey guys, I want to start a new AR Flutter project. I need to create a little MVP in two weeks for my client so I can't just wait the @CariusLars's plugin to be finished until there hehe.
> 
> You guys recommend using ARKit and ARCore plugins in the same project? Does that consumes a lot of storage capacity? Anyone have a project to share, just to understand how to manage both plugins inside the same project? I'll be very glad.
> 
> I've seen a lot of people using https://pub.dev/packages/flutter_unity_widget to create simple AR apps, but I don't know how to use the Unity... so I'm a bit afraid. Does the AR on the Unity Plugin compatible for IOS and Android??
> 
> When the ar_flutter_plugin is ready, I will be, certainly using it!

Hey @Marionetista ,
unfortunately, I don't have any code to share for you to get started, but here a few thoughts on the topic that might help you:
- First off, if you create a MVP, why go cross platform? If your client doesn't specifically ask for it, I think presenting a MVP that works on one platform only will save you a lot of work and you can go cross-platform once you start actual development
- Using ARKit and ARCore plugin in the same project: As mentioned, I haven't done it myself, but a basic approach could be checking the target platform right at the top-level widget and then re-routing to either the ARCore or ARKit widgets. That would mean you essentially have 2 completely separate apps, but if cross-platform is necessary and you only have 2 weeks that could be an easy workaround
- Regarding storage capacity: All flutter apps are quite storage-heavy, so I think the plugins wouldn't tip the scales too much (again, haven't tested it myself so might be worth just trying it out). Just keep in mind to minimize the use of compile-time assets (3D objects, images, etc.), these can blow up storage quickly and oftentimes can be queried from the web on-demand
- Regarding the unity plugin: I've also heard people use unity inside flutter and it seems to work well. Unity itself certainly supports both platforms, but I have no idea how hard it is to connect everything up in the Flutter app. If you find a sample app that's similar to what you want to do, I'd say go for it, otherwise it might be hard to get used to unity within the 2 weeks you go

I hope that helps somehow!

**Comment by RobinLbt** [comment]
@CariusLars I've seen you are making a lot of progress on AR_plugin !
Where can I pm you to help you out with the geo-based location ? I've almost finished all the functions for calculation, I would like to give you some help

**Comment by CariusLars** [comment]
> @CariusLars I've seen you are making a lot of progress on AR_plugin !
> Where can I pm you to help you out with the geo-based location ? I've almost finished all the functions for calculation, I would like to give you some help

Thanks @RobinLbt ! Just write me an email at carius.lars@gmail.com, looking forward!

**Comment by CariusLars** [comment]
Hey everyone, the plugin is finally available! 🚀 Check it out on https://pub.dev/packages/ar_flutter_plugin and contribute at https://github.com/CariusLars/ar_flutter_plugin! 💪

**Comment by guleranaa0179** [comment]
Hey guys, we want to make a cross platform AR application. we need to create a AR based menu. we are too confused in that.. should we have to use flutter/React native.
because so far we've searched about both platforms.. about flutter we found that there is no exact official plugin to implement both Arkit and ARcore in one flutter application.
an don't have much idea about react native.

You guys recommend using ARKit and ARCore plugins in the same project? we have no idea how just to manage both plugins inside the same project? if anyone have idea kindly guide us through this...

**Comment by guleranaa0179** [comment]
Also  
@CariusLars share a link of a plugin ar_flutter_plugin .
but how we know it's use. because there no any proper guidance/tutorial available.
@CariusLars can you please guide us?

**Comment by CariusLars** [comment]
A good first step would be looking at the example applications in the plugin. That should help understand how to use the plugin and also see its limitations. You can then go further into the code and see how the plugin wraps around ARCore and ARKit and if you need additional functionality, fork the repo and include it :)

**Comment by guleranaa0179** [comment]
Thanks for your response @CariusLars.
can you please share your opinion about.. should we go for flutter or React Native for this application(which i mentioned above)?
we have 3-4 months to complete this project.

**Comment by CariusLars** [comment]
Hey @guleranaa0179 , that depends completely on your prior experience. I haven't worked with React Native so I can't really say anything about it, but you can use ViroReact as an AR Framework if it supports the features you need (it's been abandoned from the official side for a while but there's still a community around it). If you have a budget, Wikitude is probably a good choice for someone without much prior experience, but it's expensive.
Anyway I'd go with the framework your more comfortable with and then have a look which AR plugin suits your needs best

**Comment by busterbogheart** [comment]
@guleranaa0179 , I just built an AR app with no mobile experience in 3 months with RN and [DeepAR.ai](https://DeepAR.ai).  So you can do it in much less time!  DeepAR is not free, but seems reasonable, and is very performant.

**Comment by guleranaa0179** [comment]
thanks for your input @busterbogheart 
it means there is a possible way to integrate it for both ios and android.

**Comment by guleranaa0179** [comment]
@busterbogheart
can you please guide us about its integration process?
that how can we integrate it in one app?
or if you about any guidance then plz recommend

**Comment by busterbogheart** [comment]
The DeepAR website has example projects showing RN integration.  They only need a little updating for the newer SDKs of DeepAR.

**Comment by gladson** [comment]
Guys take a look at this project.  Google itself started it.

https://google.github.io/filament/

https://github.com/google/filament

**Comment by rudsberg** [comment]
I am currently analyzing if Flutter may be a suitable cross-platform solution for AR apps. I have one question for people developing using ARCore/ARKit with Platform views: how is the overall performance and AR experience? Same as that to a native-built AR app? I've read [here](https://docs.flutter.dev/development/platform-integration/platform-views#performance), from the Flutter website, that there may be performance reductions due to the way the threads the Platform views are rendered. For Flutter to be seen as a viable solution, the performance must be top-notch (not noticeably worse than a native equivalent). I would love to hear about others' experiences on this. @CariusLars @khalithartmann @neokree

**Comment by RobinLbt** [comment]
Hey there @rudsberg , last time I've checked when making a production app on both platforms using Flutter, it was working really well on iOS, but very poorly on Android

But then again, I was using android AR flutter package, which is not really well maintained

@CariusLars plugin was great last time I've checked, but I didn't tried it in prod

**Comment by mfaizanfalcon** [comment]
any plans for it @Flutter Team?

**Comment by jonahwilliams** [comment]
We have no plans to add ARKit/ArCore support to the engine or to provide a first party plugin
